src/sidebar/components/filter-status.js

```diff
@@ -0,0 +1,153 @@
+ */
+export default function FilterStatus() {
+  const rootThread = useRootThread();
+
+  const filterState = useStore(store => store.filterState());
+
+  // The total count of all of the annotations in the store—may differ from
+  // the number of annotations in the thread and the number of visible annotations
+  const totalCount = useStore(store => store.annotationCount());
+
+  // Actions
+  const clearSelection = useStore(store => store.clearSelection);
+  const toggleFocusMode = useStore(store => store.toggleFocusMode);
+
+  const filterMode = (state => {
+    if (state.selectedCount > 0) {
+      return 'selection';
+    } else if (state.filterQuery) {
+      return 'query';
+    } else if (state.focusConfigured) {
+      return 'focus';
+    }
+    return null;
+  })(filterState);
+
+  if (!filterMode) {
+    // Nothing to do here
+    return null;
+  }
+
+  // Some threads in the `visibleCount` may have been "forced visible" by
+  // the user by clicking "Show x more in conversation" — subtract these
+  // forced-visible threads out to get a correct count of actual filter matches.
+  // In 'selection' mode, rely on the count of selected annotations
+  const visibleCount = countVisible(rootThread);
+  const resultCount =
+    filterMode === 'selection'
+      ? filterState.selectedCount
+      : visibleCount - filterState.forcedVisibleCount;
+
+  let buttonText;
+  switch (filterMode) {
+    case 'selection':
+      // In user-focus mode, displayed counts include annotations AND replies,
+      // while `totalCount` only includes top-level annotations, so that count
+      // doesn't make sense in user-focus mode
+      buttonText = filterState.focusConfigured
+        ? 'Show all'
+        : `Show all (${totalCount})`;
+      break;
+    case 'focus':
+      if (!filterState.forcedVisibleCount) {
+        buttonText = filterState.focusActive
+          ? 'Show all'
+          : `Show only ${filterState.focusDisplayName}`;
+      } else {
+        // When user focus is applied and there are some forced-visible threads,
+        // this special case applies. The button will clear force-visible threads
+        // but leave user focus intact
+        buttonText = 'Reset filters';
+      }
+      break;
+    case 'query':
+    default:
+      buttonText = 'Clear search';
+      break;
+  }
+
+  const buttonProps = {
+    buttonText,
+    onClick: () => clearSelection(),
+    icon: filterMode !== 'focus' ? 'cancel' : null,
+  };
+
+  // In most cases, the action button will clear the current (filter) selection,
+  // but when in 'focus' mode alone, it will toggle activation of the focus
+  if (filterMode === 'focus' && !filterState.forcedVisibleCount) {
+    buttonProps.onClick = () => toggleFocusMode();
+  }
+
+  return (
+    <div className="filter-status">
+      <div className="u-layout-row--align-center">
+        <div className="filter-status__text">
+          {visibleCount > 0 && <span>Showing</span>}{' '}
```

Comment:
I see what you mean about the text construction getting involved in the sense that there are lots of conditionals here. I don't immediately have a suggestion for that, other than to make sure you have a good set of tests which check that the resulting text (ignoring formatting) is what is expected under the various circumstances.

Reply:
I refactored this whole component a few times to see how much I could whittle it down and simplify it. I'm open to any and all suggestions!